tracks/csharp/exercises/resistor-color/mentoring.md

```diff
@@ -0,0 +1,97 @@
+### Reasonable solutions
+#### Array Based
+``` 
+public static class ResistorColor
+{
+    private static readonly string[] resistorColors = { "black", "brown", "red", "orange", "yellow", "green", "blue", "violet", "grey", "white" };
+
+    public static int ColorCode(string color) => System.Array.IndexOf(resistorColors, color);
+
+    public static string[] Colors() => resistorColors;
+}
+```
+#### Dictionary Based
+Based on Horatiu's solution
+```
+﻿using System.Collections.Generic;
+ using System.Collections.ObjectModel;
+ using System.Linq;
+  
+ public static class ResistorColor
+ {
+     private static ReadOnlyDictionary<string, int> colors = new ReadOnlyDictionary<string, int>(new Dictionary<string, int>
+     {
+         {"black", 0},
+         {"brown", 1},
+         {"red" , 2},
+         {"orange", 3},
+         {"yellow", 4},
+         {"green", 5},
+         {"blue", 6},
+         {"violet", 7},
+         {"grey", 8},
+         {"white", 9}
+     });
+ 
+     public static int ColorCode(string color)
+     {
+         return colors[color];
+     }
+ 
+     public static string[] Colors()
+     {
+         return colors.Keys.ToArray();
+     }
+ }
+```
+#### Enum Based
+```
+using System;
+
+public static class ResistorColor
+{
+    enum Color
+    {
+        black,
+        brown,
+        red,
+        orange,
+        yellow,
+        green,
+        blue,
+        violet,
+        grey,
+        white
+        
+    }
+    public static int ColorCode(string color)
+    {
+        return (int)Enum.Parse(typeof(Color), color);
+    }
+
+    public static string[] Colors()
+    {
+        return Enum.GetNames(typeof(Color));        
+    }
+}
+```
+### Common suggestions
+* Enum version: If users will be inputting or reading-out number values for the
+resistors than it might be more expressive to assign the numbers to
+the enums to show that the actual values are important. 
+* Array version: consider cloning the array on return from `Colors()` as
+a defensive measure.  This will prevent a user of the class from
+accidentally overwriting the values.
+### Talking points
+* This may or may not be the place to handle the letter case of the colors.
+* Array version: students are frequently calling `Colors()` from
+`ColorCode()` and newing the array there each time.  This does not 
+give optimal performance.
+* Dictionary version: The returned dictionary should be read only as 
+a defensive measure.  This will prevent a user of the class from
+accidentally overwriting the values.
+* Arguably the most maintainable solution would be a dictionary mapping
+string color names to explicit enums.  This would apply if both color
+names and color values had a role outside of the program.  `enum`s could 
```

Comment:
I do mean "outside the program".  In particular might I see the numeric values for resistors mentioned in business rules.  The intention is to avoid traps for the maintainer such as arise from implicit mappings.  I think it goes without saying that the values have a role outside the class but if this is all a bit too theoretical or reflects an idiosyncratic point of view I am happy to withdraw it.  Let me know.

Reply:
I think that's a great point, but the `outside of the program` line is too concise to convey it. In this case I would write it out. Something like:

```
This would apply if both color names and color values are mapped outside of this code, for example in a design document or a spec. Implicit mappings are less maintainable ánd as a bonus it's easier to mentally map the dictionary to the spec file.
```

Reply:
Fixed